src/smtencoding/FStar.SMTEncoding.EncodeTerm.fs

```diff
@@ -101,6 +101,13 @@ let whnf env t =
 let norm env t = N.normalize [Env.Beta; Env.Exclude Env.Zeta;  //we don't know if it will terminate, so no recursion
                               Env.Eager_unfolding; Env.EraseUniverses] env.tcenv t
 
+let maybe_whnf env t =
```

Comment:
I understood the other changes, but don't really get why smt encoding needs to change. Can you please add some comments to the code on how are these related to the strict arguments requirement?

Reply:
Added this comment:

```
(* `maybe_whnf env t` attempts to reduce t to weak-head normal form.
 *  It is called when `t` is a head redex, e.g., if its head symbol is marked for unfolding.
 *  However, if its head symbol is also marked as `strict_on_arguments`, then if it is applied
 *  to non-constant arguments, then it may actually not be unfolded.
 *  In those cases `maybe_whnf env t` may not reduce `t` at all.
 *  In callers of this code, we need to be careful to check that if `t` was not reduced, then
 *  we do not enter into infinite loops by recursing on `t` itself.
 *)
```